clippy_lints/src/literal_string_with_formatting_arg.rs

```diff
@@ -0,0 +1,101 @@
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let x: Option<usize> = None;
+    /// let y = "hello";
+    /// x.expect("{y:?}");
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let x: Option<usize> = None;
+    /// let y = "hello";
+    /// x.expect(&format!("{y:?}"));
+    /// ```
+    #[clippy::version = "1.83.0"]
+    pub LITERAL_STRING_WITH_FORMATTING_ARG,
+    suspicious,
+    "Checks if string literals have formatting arguments"
+}
+
+declare_lint_pass!(LiteralStringWithFormattingArg => [LITERAL_STRING_WITH_FORMATTING_ARG]);
+
+impl EarlyLintPass for LiteralStringWithFormattingArg {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if let ExprKind::Lit(lit) = expr.kind {
+            let add = match lit.kind {
+                LitKind::Str => 1,
+                LitKind::StrRaw(nb) => nb as usize + 2,
+                _ => return,
+            };
+            let fmt_str = lit.symbol.as_str();
+            let lo = expr.span.lo();
+            let mut current = fmt_str;
+            let mut diff_len = 0;
+
+            let mut parser = Parser::new(current, None, None, false, ParseMode::Format);
+            let mut spans = Vec::new();
+            while let Some(piece) = parser.next() {
+                if let Some(error) = parser.errors.last() {
+                    // We simply ignore the errors and move after them.
+                    current = &current[error.span.end + 1..];
+                    diff_len = fmt_str.len() - current.len();
+                    parser = Parser::new(current, None, None, false, ParseMode::Format);
+                } else if let Piece::NextArgument(arg) = piece {
+                    let mut pos = arg.position_span;
+                    pos.start += diff_len;
+                    pos.end += diff_len;
+
+                    let start = fmt_str[..pos.start].rfind('{').unwrap_or(pos.start);
+                    // If this is a unicode character escape, we don't want to lint.
+                    if start > 1 && fmt_str[start - 2..].starts_with("\\u{") {
+                        continue;
+                    }
+
+                    let mut end = fmt_str[pos.end..]
+                        .find('}')
+                        .map(|found| found + pos.end)
+                        .unwrap_or(pos.end);
+                    if fmt_str[start..end].contains(':') {
+                        end += 1;
+                    }
+                    spans.push(
+                        expr.span
+                            .with_hi(lo + BytePos((start + add) as _))
+                            .with_lo(lo + BytePos((end + add) as _)),
+                    );
+                }
+            }
+            if spans.len() == 1 {
+                span_lint(
+                    cx,
+                    LITERAL_STRING_WITH_FORMATTING_ARG,
+                    spans,
+                    "this is a formatting argument but it is not part of a formatting macro",
```

Comment:
Maybe a less strong wording would be good, e.g. "this looks like a formatting argument…" ? Just in case of false positives